ui/dashboards/src/components/Variables/Variable.tsx

```diff
@@ -34,33 +41,53 @@ export function TemplateVariable({ name }: TemplateVariableProps) {
   return <div>Unsupported Variable Kind: ${kind}</div>;
 }
 
+/**
+ * Returns a serialized string of the current state of variable values.
+ */
+function getVariableValuesKey(v: VariableStateMap) {
+  return Object.values(v)
+    .map((v) => JSON.stringify(v.value))
+    .join(',');
+}
+
 function ListVariable({ name }: TemplateVariableProps) {
   const ctx = useTemplateVariable(name);
   const definition = ctx.definition as ListVariableDefinition;
   const { data: variablePlugin } = usePlugin('Variable', definition.spec.plugin.kind);
-
   const { setVariableValue, setVariableLoading, setVariableOptions } = useTemplateVariableActions();
   const datasources = useLegacyDatasources();
+
+  let dependsOnVariables: string[] | undefined;
+  if (variablePlugin?.dependsOn) {
+    dependsOnVariables = variablePlugin.dependsOn(definition);
+  }
+
+  const variables = useTemplateVariableValues(dependsOnVariables);
   const allowMultiple = definition?.spec.allowMultiple === true;
   const allowAllValue = definition?.spec.allowAllValue === true;
 
-  const loadOptions = useCallback(async () => {
-    if (!variablePlugin) {
-      return;
-    }
-    setVariableLoading(name, true);
-    try {
-      const { data } = await variablePlugin.getVariableOptions(definition, { datasources });
-      setVariableOptions(name, data);
-    } catch (e) {
-      console.error('Failed to load template variable options', e);
-    }
-    setVariableLoading(name, false);
-  }, [variablePlugin, definition, name, setVariableLoading, setVariableOptions, datasources]);
+  let waitToLoad = false;
+  if (dependsOnVariables) {
+    waitToLoad = dependsOnVariables.some((v) => variables[v]?.loading);
+  }
+
+  const variablesValueKey = getVariableValuesKey(variables);
+
+  const variablesOptionsQuery = useQuery(
+    [name, definition, variablesValueKey],
+    async () => {
+      const resp = await variablePlugin?.getVariableOptions(definition, { datasources, variables });
+      return resp?.data ?? [];
+    },
+    { enabled: !!variablePlugin || waitToLoad }
+  );
 
   useEffect(() => {
-    loadOptions();
-  }, [loadOptions]);
+    setVariableLoading(name, variablesOptionsQuery.isFetching);
```

Comment:
In a follow up, I'm thinking about moving all this logic to the TemplateVariableStore so the component is more lightweight